### PR TITLE
Initial pass at the cookbook

### LINKED ISSUE #471: Some mechanism where users can share interesting sessions
**How would this feature be useful?**

I'm willing to bet that there's quite a few nox users out there who are doing weird and wonderful things with their nox sessions, outside the more common sessions like run pytest, linting etc.

I think it would be a great learning resource if nox had some way of users sharing these novel sessions in some sort of "semi-official" way (i.e. not just in passing on a related issue etc.).

When I was first starting with nox and wanting to do more complex things that aren't listed in the docs (for good reason, you don't want main project docs cluttered with complex workflows), I took to finding random examples in different repos, and using nox's own noxfile as a guide.

**Describe the solution you'd like**

Some sort of central location (could be a separate repo in the style of the 'awesome-x' series of repos, some GitHub mechanism hosted here like discussions or something, or a section in the project docs) where users can contribute their weird and wonderful nox sessions. 

Think of it like a nox cookbook, a semi-official place where nox users can show their cool "recipes".

To be totally clear, I don't think it would be valuable to have this "cookbook" filled with "here's how you run pytest", the docs already cover this super well. This is more for users composing novel automations with nox.

I'd be happy to kick this off and contribute some of the more interesting sessions I've come across to get it started.

For example, I know [pipx](https://github.com/pypa/pipx/blob/master/noxfile.py) use their noxfile to interact with git in the process of releasing a new version. I've stolen this from them for my projects because it's awesome!

I've also set up a "dev" session for my noxfile projects that automates the entire local dev setup for a python project (creating a project-level virtualenv, installing the project and its dependencies, even configuring VSCode automatically).

I bet some people are doing much crazier things with their noxfiles! It would be great to collect some of these examples in a central place

**Describe alternatives you've considered**

I guess the alternative is just pointing to interesting use cases

### PR DESCRIPTION
**Pull request by FollowTheProcess**
Closes #471 

This is my initial scaffold for the "cookbook" talked about in the linked issue. I added two of my more "interesting" nox sessions I use regularly to kind of kick it off.

Happy to go any way with this really. If we want it in the docs that's cool, or maybe break out into something separate. Just kicking off a potential idea.

Let me know if we want any changes.

Thanks 👍🏻

### CODE REVIEW

**Comment by DiddiLeija** [comment]
I **love** the idea @FollowTheProcess! I would like to keep this on the docs, but maybe we should ask another maintainer, like @theacodes.

**Comment by henryiii** [comment]
This looks really nice! Here are a couple of simpler recipes I use a lot. Maybe recipes could be collection in GitHub Discussions Show & Tell, then the best added?


```python
# Sphinx
@nox.session(reuse_venv=True)
def docs(session):
    """
    Build the docs. Pass "serve" to serve.
    """

    session.install("-e", ".[docs]")
    session.chdir('docs')

    if "pdf" in session.posargs:
        session.run("sphinx-build", "-M", "latexpdf", ".", "_build")
        return

    session.run("sphinx-build", "-M", "html", ".", "_build")

    if "serve" in session.posargs:
        session.log("Launching docs at http://localhost:8000/ - use Ctrl-C to quit")
        session.run("python", "-m", "http.server", "8000", "-d", "_build/html")
    elif session.posargs:
        session.error("Unsupported argument to docs")
            
# MkDocs
@nox.session(reuse_venv=True)
def docs(session: nox.Session) -> None:
    """
    Build the docs.
    """
    session.install("-e", ".[docs]")

    if session.posargs:
        if "serve" in session.posargs:
            session.run("mkdocs", "serve")
        else:
            session.error("Unrecognized args, use 'serve'")
    else:
        session.run("mkdocs", "build")
```

```python
@nox.session(reuse_venv=True)
def lint(session):
    """
    Run the linter.
    """
    session.install('pre-commit')
    session.run('pre-commit', 'run', '--all-files', *session.posargs)
```

```python
@nox.session(reuse_venv=True)
def build(session):
    """
    Build an SDist and wheel.
    """

    session.install('build')
    session.run('python', '-m', 'build', *session.posargs)
```

File: docs/cookbook.rst [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+The What?
+---------
+
+A lot of people and a lot of projects use ``nox`` for their python automation powers.
+
+Some of these sessions are the classic "run pytest and linting", some are more unique and more interesting!
+
+The nox cookbook is a collection of these such sessions.
+
+``nox`` is super easy to get started with, and super powerful right out of the box. But when things get complex or you want to chain together some more powerful tasks, often the only examples can be found hunting around GitHub for novel ``nox`` sessions.
```
Comment (theacodes) [inline docs/cookbook.rst:13]: Nit: use "Nox" when referring to the project, only use ``nox`` when referring to the command line program. Also, Nox is capitalized (unlike tox).

File: docs/cookbook.rst [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+The kind of sessions that make you think "I didn't know you could do that!"
+
+This cookbook is intended to be a centralized, community-driven repository of awesome ``nox`` sessions to act as a source of inspiration and a reference guide for nox's users. If you're doing something cool with ``nox``, why not add your session here?
+
+
+Contributing a Session
```
Comment (theacodes) [inline docs/cookbook.rst:20]: Maybe link to our contributing guide?

File: docs/cookbook.rst [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+----------------------
+
+Anyone can contribute sessions to the cookbook. However, there are a few guiding principles you should keep in mind:
+
+* Your session should be interesting or unique, it should do something out of the ordinary or otherwise interesting.
+* You should explain briefly what it does and why it's interesting.
+* You should include the full session as a code block, if it's too large, try to include a minimal example. Bonus points if a user can copy and paste it into their ``noxfile.py`` and run it straight away!
+
+
+And now on with the recipes...
```
Comment (theacodes) [inline docs/cookbook.rst:30]: You can remove this.

File: docs/cookbook.rst [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+.. code-block:: python
+
+    import os
+    
+    import nox
+
+    # It's a good idea to keep your dev session out of the default list
+    # so it's not run twice accidentally
+    nox.options.sessions = [<sessions other than dev>]
```
Comment (theacodes) [inline docs/cookbook.rst:50]: You can use an ellipsis here and it will remain valid python code.

File: docs/cookbook.rst [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        """
+        Kicks off an automated release process by creating and pushing a new tag.
+
+        Invokes bump2version with the posarg setting the version.
+        
+        Usage:
+        $ nox -s release -- [major|minor|patch]
+        """
+
+        allowed_args = {"major", "minor", "patch"}
```
Comment (theacodes) [inline docs/cookbook.rst:101]: This is a lot of code to handle argument parsing, maybe use `argparse`?

File: docs/cookbook.rst (hunk 4 above)
Comment (FollowTheProcess) [reply docs/cookbook.rst]: Always wondered what ellipsis was used for! thanks!

File: docs/cookbook.rst (hunk 5 above)
Comment (FollowTheProcess) [reply docs/cookbook.rst]: Well TIL! Not sure why I didn't think of this but this is great, I've changed all my release sessions. Thanks! 🎉

**Comment by DiddiLeija** [comment]
@FollowTheProcess, can you please update this branch, so we can go ahead with the PR?

**Review by DiddiLeija — APPROVED** [review]
I'll wait a little before merging, to see if someone has concerns about this :)

**Comment by FollowTheProcess** [comment]
I'm gonna go ahead and merge this in once the branch has updated, it's been hanging around for a while 🙂

**Pull request merged**